### PR TITLE
Preserve existing keys in dictionary

### PR DESCRIPTION
**Pull request by mapogolions**
At the moment, binding to dictionary has two different behaviors depending on the nesting level.

1) Behaviour 1 - add | override | preserve. Please follow the links below for each case 
[add new & preserve existing keys](https://github.com/dotnet/runtime/blob/c30d55ec5b584e26186a85adedf313c81a64269c/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs#L470)
[override existing keys](https://github.com/dotnet/runtime/blob/c30d55ec5b584e26186a85adedf313c81a64269c/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs#L496)

2) Behaviour 2 - skip all existing keys & add new
Meaning that new keys from a configuration will be picked up, but already existing dropped
Please see the commit that reproduces the issue and shows in which cases this happens. 

This PR attempts to correct this inconsistent behavior and use the first approach as baseline.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 1]
```diff
@@ -457,27 +457,21 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                 // We only support string and enum keys
                 return;
             }
-
+            MethodInfo tryGetValue = dictionaryType.GetMethod("TryGetValue")!;
             PropertyInfo setter = dictionaryType.GetProperty("Item", DeclaredOnlyLookup)!;
             foreach (IConfigurationSection child in config.GetChildren())
             {
+                var key = keyTypeIsEnum ? Enum.Parse(keyType, child.Key) : child.Key;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:464]:
```suggestion
                object key = keyTypeIsEnum ? Enum.Parse(keyType, child.Key) : child.Key;
```
We don't use `var` unless it is obvious from the right hand side.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 2]
```diff
@@ -457,27 +457,21 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
...
+                var args = new object?[] { key, null };
+                var _ = tryGetValue.Invoke(dictionary, args);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:466]:
```suggestion
                _ = tryGetValue.Invoke(dictionary, args);
```

The difference is your code is declaring a variable named `_`. With the suggestion we are explicitly telling the compiler we are ignoring the result.

**Comment by mapogolions** [comment]
@eerhardt Thanks for review. I've added a unit test for a dictionary with an enum as key type

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs [hunk 3]
```diff
@@ -466,6 +466,55 @@ public void StringDictionaryBinding()
             Assert.Equal("val_3", options.StringDictionary["ghi"]);
         }
 
+        [Fact]
```
[the remark this answers is not in the document]
Comment (mapogolions) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs:469]: Please see the comment below

**Review by eerhardt — APPROVED** [review]
The changes look good to me. Thanks for the contribution!

@maryamariyan - any thoughts?

**Comment by eerhardt** [comment]
OSX Failure is the same listed in https://github.com/dotnet/runtime/issues/63439#issuecomment-1014912407.

**Comment by maryamariyan** [comment]
There's a conflict on the ConfigurationBinder file left to resolve. I think happened after merging your other PR.

Since Microsoft.Extensions.Configuration is widely used we're always very careful with taking risk on fixing/changing behaviors in it in case there are clients who already rely on the existing behavior, even if it is not correct. 

For the very least we'd need to document the breaking change. @mapogolions do you see any use cases where we'd have regressing behavior?

cc @HaoK

**Comment by mapogolions** [comment]
@maryamariyan 
The conflict was probably caused by [this PR](https://github.com/dotnet/runtime/pull/63792) that wraps each iteration in a try-catch block. In my opinion, this PR should not affect existing code.
Old approach:
`drop existing keys & add new keys == (override if key exists | add key if not found in the origin) & remove all keys that are unique only to the origin`
New approach:
`(override if key exists | add key if not found in the origin) & do nothing`
As you can see, only the tail has been changed

**Review by maryamariyan — APPROVED** [review]
LGTM aside from the merge conflict that needs to be resolved

**Comment by eerhardt** [comment]
Failure is unrelated. Merging.

**Pull request merged**